### PR TITLE
refactor(api)!: json data for ContentNode (new approach)

### PR DESCRIPTION
**Pull request by usu**
This is an alternative POC for leveraging json data:
- keeps ContentNode subclasses, but changes to Single Table Inheritance (thus avoiding all the extra joins for every subclass)
- still utilizes $data JSONB property for most use cases (but additional data could be linked for more complex cases)
- no switch statements anymore 😎 

Replaces the POC discussed under https://github.com/usu/ecamp3/pull/8/commits

### Open ToDos to make tis POC mergeable
- [x] fix tests
- [x] fix frontend
- [x] fix printing
- [x] migrate dev data (or start from scratch?)
- [x] implement copyFromPrototype
- [x] verify Swagger output
- [x] fix frontend tests

### To discuss
- [x] keep reference to materialItems from materialNode? (I tend to pay the small performance impact and say yes to simplify coding)

### To do (separate PR?)
- [ ] unit tests JsonMergePatch #2944
- [ ] ApiWrapper: consider renaming 'fieldname' to 'path', as this now allows to contain a nested json path #2945

### Time measurement
(contentNodes timing not fully comparable, as data is not migrated)

Endpoint | Collection ms | Entity ms
-- | -- | --
activities | 240 | 114
contentNodes | 232 | 73
columnLayouts | 126 | 74
multiSelects | 70 | 68
singleTexts | 112 | 70
storyboards | 88 | 69
materialNodes | 82 | 71
days | 82 | 68
categories | 102 | 83
contentTypes | 58 | 52
camps | 84 | 83
periods | 87 | 85
materialLists | 74 | 62
scheduleEntries | 102 | 81
materialItems | 90 | 70
activityResponsibles | 65 | 65
dayResponsibles | 61 | 64
campCollaborations | 103 | 70

### CODE REVIEW

File: api/src/Entity/ContentNode.php [hunk 1]
```diff
@@ -95,6 +81,15 @@ abstract class ContentNode extends BaseEntity implements BelongsToCampInterface,
     #[ORM\OneToMany(targetEntity: ContentNode::class, mappedBy: 'parent', cascade: ['persist'])]
     public Collection $children;
 
+    /**
+     * Holds the actual data of the content node.
+     */
+    #[ApiProperty(example: ['text' => 'dummy text'])]
+    #[Groups(['read', 'write'])]
+    #[ORM\Column(type: 'json', nullable: true, options: ['jsonb' => true])]
+    #[Assert\IsNull(groups: ['create'])] // create with empty data; default value is populated by ContentNodeDataPersister
```
Comment (BacLuc) [inline api/src/Entity/ContentNode.php]: It is our ui which does not allow to create a contentnode directly with data.
We could allow that for other ui's.

Did you test how the Assertions are merged in subclasses on overridden properties?

File: api/src/Entity/HasRootContentNodeTrait.php [hunk 2]
```diff
@@ -3,50 +3,47 @@
...
+use App\Serializer\Normalizer\RelatedCollectionLink;
 use Doctrine\ORM\Mapping as ORM;
+use Symfony\Component\Serializer\Annotation\Groups;
 use Symfony\Component\Validator\Constraints as Assert;
 
-#[ORM\Entity]
-#[ORM\InheritanceType('JOINED')]
-#[ORM\DiscriminatorColumn(name: 'entityType', type: 'string')]
-abstract class AbstractContentNodeOwner extends BaseEntity {
+trait HasRootContentNodeTrait {
```
Comment (BacLuc) [inline api/src/Entity/HasRootContentNodeTrait.php]: minor: a little inconsistent that the trait is HasRootContentNodeTrait, but for the root we only allow ColumnLayout.
I am not sure if we should name the trait HasColumnLayoutTrait.

File: api/src/Entity/Period.php [hunk 3]
```diff
@@ -279,16 +277,20 @@ public function getPeriodLength(): ?int {
...
+            }
+
+            if ($period->start > $this->start) {
+                return false;
+            }
 
-        /** @var Selectable $periodCollection */
-        $periodCollection = $this->camp->periods;
-        $periods = $periodCollection->matching($crit);
+            return $period->getId() < $this->getId();
```
Comment (BacLuc) [inline api/src/Entity/Period.php]: why do you compare the period id?
Isn't that a random generated string?

File: api/src/Repository/FiltersByContentNode.php [hunk 4]
```diff
@@ -21,18 +21,15 @@ trait FiltersByContentNode {
...
 
         /*
-         * COALESCE:
-         *   If owner is an Activity --> cn_activity_category.id is properly loaded (not null)
-         *   If owner is a Category --> cn_activity_category.id is null --> category is loaded via owner.id
+         *   If owner is an Activity --> cn_activity.category is not null
+         *   If owner is a Category --> cn_category.rootContentNode = root.id is a match
          */
-        $queryBuilder->join(Category::class, 'cn_category', Join::WITH, 'cn_category.id = COALESCE(cn_activity_category.id, owner.id)');
+        $queryBuilder->join(Category::class, 'cn_category', Join::WITH, 'cn_category.id = cn_activity.category OR cn_category.rootContentNode = root.id');
```
Comment (BacLuc) [inline api/src/Repository/FiltersByContentNode.php]: Here is 2 times a comparison between an id and an object.
I would use the object on both sides.
cn_category.id = cn_activity.category

File: api/src/Util/JsonMergePatch.php [hunk 5]
```diff
@@ -0,0 +1,47 @@
+<?php
+
+namespace App\Util;
+
+class JsonMergePatch {
+    /**
+     * implements JSON merge patch (https://datatracker.ietf.org/doc/html/rfc7386)
+     * partially copied by discussion under https://www.php.net/manual/de/function.array-merge-recursive.php.
+     */
+    public static function mergePatch(array &$target, array &$patch) {
```
Comment (BacLuc) [inline api/src/Util/JsonMergePatch.php]: Do we need to change the target in place?
(We for sure don't want to change the $patch in place)

I would prefer it if we returned the merged array instead of in place modification.

File: api/src/Util/JsonMergePatch.php [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+            // associative arrays --> merge recursively
+            if (is_array($value) && self::isAssoc($value) && isset($merged[$key]) && is_array($merged[$key])) {
+                $merged[$key] = self::mergePatch($merged[$key], $value);
+
+                continue;
+            }
+
+            // primitve values or sequential/numeric arrays (true JSON arrays)
+            $merged[$key] = $value; // TODO: Consider/discuss to apply CleanHTMLFilter cirectly here for all string values
```
Comment (BacLuc) [inline api/src/Util/JsonMergePatch.php:29]: sequential arrays may still contain arrays as values

File: api/src/Util/JsonMergePatch.php [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+        foreach ($patch as $key => &$value) {
+            // null values can be used to remove keys/elements from the structure
+            if (is_null($value)) {
+                unset($merged[$key]);
+
+                continue;
+            }
+
+            // associative arrays --> merge recursively
+            if (is_array($value) && self::isAssoc($value) && isset($merged[$key]) && is_array($merged[$key])) {
```
Comment (BacLuc) [inline api/src/Util/JsonMergePatch.php]: Maybe use https://wiki.php.net/rfc/is_list instead of self::isAssoc

Comment (usu) [reply api/src/Util/JsonMergePatch.php]: Ah nice, didn't know this exists

**Comment by usu** [comment]
PR would we ready for review. @pmattmann & @carlobeltrame, do you have a chance to look at it before the meeting this week?

File: api/src/Entity/ContentNode.php (hunk 1 above)
Comment (usu) [reply api/src/Entity/ContentNode.php]: > It is our ui which does not allow to create a contentnode directly with data.
> We could allow that for other ui's.

It's now changed and it's possible to send data during create already. An exception is the MultiSelect where  I kept the 
`#[Assert\IsNull(groups: ['create'])]` and the data is populated from ContentType in the data persister.

File: api/src/Entity/Period.php (hunk 3 above)
Comment (usu) [reply api/src/Entity/Period.php]: Seems not part of this pull request. But I believe it's the same as with ScheduleEntry numbers: If all other things are equal (in the case of the period the start property), then the id is taken to provide a consistent numbering.

File: api/src/Util/JsonMergePatch.php (hunk 7 above)
Comment (usu) [reply api/src/Util/JsonMergePatch.php]: Fixed with https://github.com/ecamp/ecamp3/pull/2825/commits/75fff84060938a30957df13a350bdd36cd4d9abd

File: api/src/Util/JsonMergePatch.php (hunk 5 above)
Comment (usu) [reply api/src/Util/JsonMergePatch.php]: Fixed with https://github.com/ecamp/ecamp3/pull/2825/commits/75fff84060938a30957df13a350bdd36cd4d9abd

File: api/src/Repository/FiltersByContentNode.php (hunk 4 above)
Comment (usu) [reply api/src/Repository/FiltersByContentNode.php]: Fixed with https://github.com/ecamp/ecamp3/pull/2825/commits/75fff84060938a30957df13a350bdd36cd4d9abd

File: api/src/Util/JsonMergePatch.php (hunk 6 above)
Comment (usu) [reply api/src/Util/JsonMergePatch.php:29]: In which case the target would be overriden and not merged. To my understanding this is the expected behavior of `JSON merge patch`

File: api/migrations/schema/Version20220611193723.php [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+        $this->addSql('DROP TABLE content_node_columnlayout');
+        $this->addSql('DROP TABLE content_node_materialnode');
+        $this->addSql('DROP TABLE content_node_multiselect');
+        $this->addSql('DROP TABLE content_node_singletext');
+        $this->addSql('DROP TABLE content_node_storyboard');
+        $this->addSql('DROP TABLE content_node_multiselect_option');
+        $this->addSql('DROP TABLE content_node_storyboard_section');
+        $this->addSql('ALTER TABLE activity ADD CONSTRAINT FK_AC74095AF886581C FOREIGN KEY (rootContentNodeId) REFERENCES content_node (id) ON DELETE CASCADE NOT DEFERRABLE INITIALLY IMMEDIATE');
+        $this->addSql('ALTER TABLE category ADD CONSTRAINT FK_64C19C1F886581C FOREIGN KEY (rootContentNodeId) REFERENCES content_node (id) ON DELETE CASCADE NOT DEFERRABLE INITIALLY IMMEDIATE');
+        $this->addSql('ALTER TABLE content_node ADD CONSTRAINT FK_481D0580B7939B21 FOREIGN KEY (rootId) REFERENCES content_node (id) NOT DEFERRABLE INITIALLY IMMEDIATE');
```
Comment (carlobeltrame) [inline api/migrations/schema/Version20220611193723.php]: Müsste hier nicht auch ON DELETE CASCADE stehen?

File: api/src/DataPersister/ActivityDataPersister.php [hunk 9]
```diff
@@ -33,6 +36,11 @@ public function beforeCreate($data): Activity {
         }
 
         $rootContentNode = new ColumnLayout();
+        $rootContentNode->contentType = $this->em
+            ->getRepository(ContentType::class)
+            ->findOneBy(['name' => 'ColumnLayout'])
+        ;
+        $rootContentNode->data = ['columns' => [['slot' => '1', 'width' => 12]]];
```
Comment (carlobeltrame) [inline api/src/DataPersister/ActivityDataPersister.php]: Wo wurde das zuvor gemacht?

File: api/src/Entity/ContentNode.php [hunk 10]
```diff
@@ -149,6 +158,22 @@ public function getRoot(): ?ColumnLayout {
...
+
+    public function setData(?array $data) {
+        if (null === $this->data) {
+            $this->data = $data;
+
+            return;
+        }
+
+        if (null !== $data) {
+            $this->data = JsonMergePatch::mergePatch($this->data, $data);
```
Comment (carlobeltrame) [inline api/src/Entity/ContentNode.php:173]: Hätten wir hier irgendwie die Möglichkeit, auch mal JsonPatch zu implementieren, und dann je nach ContentType im Request eine andere Merge-Strategie zu verwenden? Bei dieser Frage bekomme ich irgendwie das Gefühl dass das hier eher in einen Denormalizer gehört. Wäre das eine Möglichkeit?

Wenn du das im Setter machst wirds auch z.B. schwieriger, Tests zu schreiben etc.

File: api/src/Entity/ContentNode/SingleText.php [hunk 11]
```diff
@@ -35,20 +35,25 @@
...
-     * @param EntityMap  $entityMap
+     * Holds the actual data of the content node
+     * (overridden from abstract class in order to add specific validation).
      */
-    public function copyFromPrototype($prototype, $entityMap): void {
-        parent::copyFromPrototype($prototype, $entityMap);
-
-        $this->text = $prototype->text;
-    }
+    #[ApiProperty(example: ['text' => 'dummy text'])]
```
Comment (carlobeltrame) [inline api/src/Entity/ContentNode/SingleText.php]:
```suggestion
    #[ApiProperty(example: ['text' => 'my example text'])]
```
Das kommt in die OpenAPI Dokumentation, daher sollte es neuen Entwickler*innen helfen zu verstehen dass das ein Freitext-Teil des Payloads ist.

File: api/src/Entity/ContentNode/MultiSelect.php [hunk 12]
```diff
@@ -37,59 +36,39 @@
...
-
-        // copy all multiSelect options
-        foreach ($prototype->options as $optionPrototype) {
-            $option = new MultiSelectOption();
-            $this->addOption($option);
-
-            $option->copyFromPrototype($optionPrototype, $entityMap);
-        }
-    }
+    #[ApiProperty(example: null)]
```
Comment (carlobeltrame) [inline api/src/Entity/ContentNode/MultiSelect.php]: Könnten wir hier trotzdem ein example einfügen? Mit dem JSON data Feld ists viel intransparenter in der Dokumentation, was überhaupt erlaubt sein könnte, und so ein Example könnte bei Get oder besonders Patch helfen.

File: api/src/Entity/ContentNode/Storyboard.php [hunk 13]
```diff
@@ -37,59 +35,65 @@
...
+        'additionalProperties' => false,
+        'required' => ['sections'],
+        'properties' => [
+            'sections' => [
+                'anyOf' => [
+                    [
+                        'type' => 'object',
+                        'patternProperties' => [
+                            // uuid4 key
+                            '[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}' => [
```
Comment (carlobeltrame) [inline api/src/Entity/ContentNode/Storyboard.php:50]: Da wir das vermutlich noch oft brauchen werden: Gibt es eine Möglichkeit, das als globale #$def irgendwo in den JSON Schema Checker einzubauen? Oder geht das nicht weils ein Key ist?

File: api/migrations/schema/Version20220611193723.php [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+        $this->addSql('ALTER TABLE content_node_storyboard_section DROP CONSTRAINT fk_e2b71b9f8f096e57');
+        $this->addSql('DROP TABLE content_node_columnlayout');
+        $this->addSql('DROP TABLE content_node_materialnode');
+        $this->addSql('DROP TABLE content_node_multiselect');
+        $this->addSql('DROP TABLE content_node_singletext');
+        $this->addSql('DROP TABLE content_node_storyboard');
+        $this->addSql('DROP TABLE content_node_multiselect_option');
+        $this->addSql('DROP TABLE content_node_storyboard_section');
+        $this->addSql('ALTER TABLE activity ADD CONSTRAINT FK_AC74095AF886581C FOREIGN KEY (rootContentNodeId) REFERENCES content_node (id) ON DELETE CASCADE NOT DEFERRABLE INITIALLY IMMEDIATE');
+        $this->addSql('ALTER TABLE category ADD CONSTRAINT FK_64C19C1F886581C FOREIGN KEY (rootContentNodeId) REFERENCES content_node (id) ON DELETE CASCADE NOT DEFERRABLE INITIALLY IMMEDIATE');
```
Comment (carlobeltrame) [inline api/migrations/schema/Version20220611193723.php]: Also wenn man den Root Content Node einer Activity oder Category löscht, dann wird jetzt auch die Activity / Category mitgelöscht..? Scheint etwas unintuitiv auf den ersten Blick, aber wir erlauben ja keine Activity oder Category ohne Content, also okay... Müssen nur aufpassen dass das keine unerwarteten Nebenwirkungen hat in Zukunft.

File: api/tests/Api/ContentNodes/ContentNode/ReadContentNodeTest.php [hunk 15]
```diff
@@ -52,20 +51,4 @@ public function testGetSingleContentNodeIsAllowedInCampPrototype() {
         // then
         $this->assertResponseStatusCodeSame(200);
     }
-
-    public function testGetSingleContentNodeIncludesProperRelationLinks() {
```
Comment (carlobeltrame) [inline api/tests/Api/ContentNodes/ContentNode/ReadContentNodeTest.php:56]: Müsste man hier einen Test als Ersatz schreiben, der überprüft dass der JSON-Inhalt rauskommt?

File: api/src/Util/JsonMergePatch.php [hunk 16]
```diff
@@ -0,0 +1,34 @@
+<?php
+
+namespace App\Util;
+
+class JsonMergePatch {
+    /**
+     * implements JSON merge patch (https://datatracker.ietf.org/doc/html/rfc7386)
+     * partially copied by discussion under https://www.php.net/manual/de/function.array-merge-recursive.php.
+     */
+    public static function mergePatch(array $target, array $patch) {
```
Comment (carlobeltrame) [inline api/src/Util/JsonMergePatch.php:10]: Ich würde mir bei selbst implementierten Algorithmen wie diesem hier ausführliche Unit Tests für alle denkbaren happy paths, sad paths und edge cases wünschen.

**Review by pmattmann — APPROVED** [review]

File: frontend/src/components/activity/content/Storyboard.vue [hunk 17]
```diff
@@ -52,26 +53,28 @@
               <v-container v-if="!layoutMode && !disabled" class="ma-0 pa-0">
                 <v-row no-gutters>
                   <v-col cols="6">
+                    <!-- TODO: dialog to ask for confirmation before deletion -->
```
Comment (carlobeltrame) [inline frontend/src/components/activity/content/Storyboard.vue:56]: Machen wir doch ein Follow-Up Issue für das, statt einem HTML-Kommentar.

File: frontend/src/components/activity/content/columnLayout/ColumnOperations.vue [hunk 18]
```diff
@@ -36,11 +36,14 @@ export default {
   computed: {
     addingColumnEnabled() {
       return (
-        (this.contentNode.columns.length + 1) * this.minColumnWidth <= this.totalWidth
+        (this.contentNode.data.columns.length + 1) * this.minColumnWidth <=
+        this.totalWidth
```
Comment (carlobeltrame) [inline frontend/src/components/activity/content/columnLayout/ColumnOperations.vue:40]: wow. very formatting. such characters :dog:

File: frontend/src/components/form/api/ApiSortable.vue [hunk 19]
```diff
@@ -24,112 +24,123 @@ import draggable from 'vuedraggable'
...
+      this.localSortedKeys.splice(indexToDelete, 1)
+
+      // generate positions
+      const payload = this.recalculatePositions()
+
+      // replace deleted item with null value
+      payload[key] = null
+
+      this.$emit('sort', payload)
+    },
```
Comment (carlobeltrame) [inline frontend/src/components/form/api/ApiSortable.vue:121]: Wir sollten wirklich beginnen, mehr Frontend-Tests für Dinge wie diese Komponente zu implementieren...

File: frontend/src/mixins/apiPropsMixin.js [hunk 20]
```diff
@@ -8,6 +8,7 @@ export const apiPropsMixin = {
     value: { required: false, default: null },
 
     /* field name and URI for saving back to API */
+    // TODO: consider renaming to 'path', as this now allows to contain a nested json path
```
Comment (carlobeltrame) [inline frontend/src/mixins/apiPropsMixin.js:11]: Ich würde eher ein Issue erstellen als einen TODO-Kommentar schreiben. Niemand sucht, kategorisiert oder updatet all die TODO-Kommentare im Code.

**Review by carlobeltrame — COMMENTED** [review]
Von nun an ist es für jeden neuen ContentNode mit Daten nötig, zusätzlich einen DataPersister zu schreiben in dem Validierungen und InputFilters ausprogrammiert werden müssen, mit einem anderen System als bei der "normalen" ContentNode Entity. Das müssen wir dann gut dokumentieren, oder vielleicht finden wir noch einen Weg das zu vereinfachen / wieder zu vereinheitlichen.
Weitere Nachteile der JSON-Daten: Es wird schwieriger werden, nur einzelne Datenpunkte auszutauschen, wenn wir mal Live-Updates einführen. Und es wird schwieriger, von ContentNode-Daten aus andere Entities zu referenzieren (Prime-Beispiel sind Kommentare, oder auch z.B. Links zu anderen Activities, wobei man dann wohl mit GUIDs und ausführlichen Fallbacks arbeiten kann...)

File: api/src/DataPersister/ActivityDataPersister.php (hunk 9 above)
Comment (usu) [reply api/src/DataPersister/ActivityDataPersister.php]: (`$rootContentNode->data = `) ist nun glaubs wieder überflüssig, weil `['columns' => [['slot' => '1', 'width' => 12]]]` wieder in der Entity als Default-Wert hinterlegt ist (hatte ich mal rausgenommen aber noch Kommentar von @BacLuc wieder rückgängig gemacht.

File: api/src/Entity/ContentNode/MultiSelect.php (hunk 12 above)
Comment (usu) [reply api/src/Entity/ContentNode/MultiSelect.php]: Yes, macht Sinn. Muss dann aber noch die Tests anpassen, weil dort wird der default-Payload anhand der example values zusammengesetzt.

File: api/src/Util/JsonMergePatch.php (hunk 16 above)
Comment (usu) [reply api/src/Util/JsonMergePatch.php:10]: Fully agree and mentally already on my list 😄 
Potentially in separate PR, let's discuss

File: frontend/src/components/activity/content/Storyboard.vue (hunk 17 above)
Comment (usu) [reply frontend/src/components/activity/content/Storyboard.vue:56]: #2933

File: api/migrations/schema/Version20220611193723.php (hunk 14 above)
Comment (usu) [reply api/migrations/schema/Version20220611193723.php]: Hmm, ja dann ists verkehrt. Löschen einer Activity/Category sollte den RootContentNode löschen (+ alle sub-nodes). Löschen eines RootContentNodes ist nicht erlaubt (gibt's auch einen Test dazu).

File: api/migrations/schema/Version20220611193723.php (hunk 8 above)
Comment (usu) [reply api/migrations/schema/Version20220611193723.php]: Im Moment ist ON DELETE CASCADE auf der parent/children Beziehung. Würde ich glaubs dort lassen, dann kann man auch Teilbäume löschen.

File: api/src/Entity/ContentNode.php (hunk 10 above)
Comment (usu) [reply api/src/Entity/ContentNode.php:173]: Idee: Serializer welcher nur bei ContentNode anspringt (#2947)

File: api/migrations/schema/Version20220611193723.php (hunk 8 above)
Comment (usu) [reply api/migrations/schema/Version20220611193723.php]: Fixed with https://github.com/ecamp/ecamp3/pull/2825/commits/f49982ed9c8b99c36436f8ebc3777e3d8713dfd2

Habens nun bei beiden Beziehungen drin, parent & root (doesn't hurt to have it at both).

File: api/migrations/schema/Version20220611193723.php (hunk 14 above)
Comment (usu) [reply api/migrations/schema/Version20220611193723.php]: Fixed with https://github.com/ecamp/ecamp3/pull/2825/commits/f49982ed9c8b99c36436f8ebc3777e3d8713dfd2
Related to #2943

File: api/src/Util/JsonMergePatch.php (hunk 16 above)
Comment (usu) [reply api/src/Util/JsonMergePatch.php:10]: Will be fixed in #2944

File: frontend/src/mixins/apiPropsMixin.js (hunk 20 above)
Comment (usu) [reply frontend/src/mixins/apiPropsMixin.js:11]: Will be fixed with #2945

File: frontend/src/components/form/api/ApiSortable.vue (hunk 19 above)
Comment (usu) [reply frontend/src/components/form/api/ApiSortable.vue:121]: #2946

File: api/src/Entity/ContentNode/Storyboard.php (hunk 13 above)
Comment (usu) [reply api/src/Entity/ContentNode/Storyboard.php:50]: #2948

File: api/src/Entity/HasRootContentNodeTrait.php (hunk 2 above)
Comment (usu) [reply api/src/Entity/HasRootContentNodeTrait.php]: Probably "Geschmackssache", but I would rather keep the naming. Even though the `rootContentNode` property is at the moment specifically a `ColumnLayout`, the role it plays in this context is to be a root content node and not to be a generic column layout content node.

`HasColumnLayoutTrait` could, purely from a name, also apply to any other `ColumnLayout` (because they can also "have" other column layouts).

File: api/src/DataPersister/ActivityDataPersister.php (hunk 9 above)
Comment (usu) [reply api/src/DataPersister/ActivityDataPersister.php]: Fixed in https://github.com/ecamp/ecamp3/pull/2825/commits/dd67ee301305c9596c196ba0562d973d13fb0c69

File: api/src/Entity/ContentNode/SingleText.php (hunk 11 above)
Comment (usu) [reply api/src/Entity/ContentNode/SingleText.php]: Fixed in https://github.com/ecamp/ecamp3/pull/2825/commits/dd67ee301305c9596c196ba0562d973d13fb0c69

File: api/src/Entity/ContentNode/MultiSelect.php (hunk 12 above)
Comment (usu) [reply api/src/Entity/ContentNode/MultiSelect.php]: Fixed in https://github.com/ecamp/ecamp3/pull/2825/commits/dd67ee301305c9596c196ba0562d973d13fb0c69

File: api/tests/Api/ContentNodes/ContentNode/ReadContentNodeTest.php (hunk 15 above)
Comment (usu) [reply api/tests/Api/ContentNodes/ContentNode/ReadContentNodeTest.php:56]: Tests extended in https://github.com/ecamp/ecamp3/pull/2825/commits/dd67ee301305c9596c196ba0562d973d13fb0c69

**Review by BacLuc — APPROVED** [review]

File: api/migrations/schema/Version20220611193723.php (hunk 14 above)
Comment (usu) [reply api/migrations/schema/Version20220611193723.php]: Detected a bug #2956
Workaround implemented in https://github.com/ecamp/ecamp3/pull/2825/commits/22231efa27c74ea042c93fd3689cd934739e6385 until bug is fixed

**Comment by usu** [comment]
@carlobeltrame Ready for delta-review. All open topics/workarounds are registered as issues.

**Review by carlobeltrame — APPROVED** [review]
